Add sort method to AtomGroup

1ut opened the pull request:
Fixes #2976

Changes made in this Pull Request:

Added sort method to the atomgroup.

- Parameters
        A function for the parameter of each atom to use for sorting. i.e. lambda atom: atom.id

- Returns
       Sorted atomgroup.



I'm new to mdanalysis so I would be grateful if you could inform me when you find any mistakes.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

pep8speaks commented:
Hello @1ut! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/core/groups.py`](https://github.com/MDAnalysis/mdanalysis/blob/f654ed5916477bca979c287e63304af5df160fc7/package/MDAnalysis/core/groups.py):

> [Line 3351:70](https://github.com/MDAnalysis/mdanalysis/blob/f654ed5916477bca979c287e63304af5df160fc7/package/MDAnalysis/core/groups.py#L3351): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

richardjgowers reviewed (changes requested):
Hi @1ut thanks for starting on this.  We'll also need a test written in test_atomgroup.py to make sure that this code is working as it should.

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,56 @@ def write(self, filename=None, file_format=None,
+                n = len(self.atoms)
+                idx = np.zeros(n)
+                for i in range(n):
+                    idx[i] = value(self.atoms[i])
```
I don't like the pattern of providing a lambda that is applied to each atom.  Mostly because it will scale poorly and AtomGroups are often large.  I think a better approach would be to take the name of an attribute (eg `"ids"`, `"masses"`) and have a `getattr` call to build the `idx` array here.  This can then still use the argsort afterwards.

1ut commented:
I've rewritten the code. I appreciate your kind suggestion.

IAlibay commented on package/CHANGELOG:
```diff
@@ -133,6 +133,7 @@ Enhancements
     'protein' selection (#2751 PR #2755)
   * Added an RDKit converter that works for any input with all hydrogens
     explicit in the topology (Issue #2468, PR #2775)
+  * Added sort method to the atomgroup (Issue #2976) 
```
Note these entries should be newest first.

Also please do add yourself to the author list for 2.0 (above) and AUTHORS.

1ut replied:
Added my name to author list for 2.0 and AUTHORS.

https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/67bef3a3e1e2d76872dfae0b5b678d4802397877
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/aeca48da8de3be1a8570168792a9c2e937ee2c7f

IAlibay reviewed (changes requested):
Couple of quick thoughts.

richardjgowers commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -1674,3 +1675,25 @@ def test_partial_timestep(self, universe):
+
+    @pytest.fixture()
+    def u(self):
+        return mda.Universe(PDB_helix)
```
generally using `Universe.empty()` is better for tests

1ut replied:
Add tests for TestAtomGroupSort for non-unique attribute using atomgroup from Universe.empty().

https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/a535599aaafcbacec22b804ced24204bda3f3157

hmacdope commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -1674,3 +1675,25 @@ def test_partial_timestep(self, universe):
+    def agsort(self, ag):
+        return ag.sort("ids")
+
+    def test_sort(self, agsort):
```
@IAlibay sorry to jump the gun a bit here, but would we want to restrict to a list of `sortable` attributes or check there is no collision (scales badly)?

richardjgowers commented:
What would be unsortable?

On Mon, Mar 29, 2021 at 12:47, Hugo MacDermott-Opeskin <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In testsuite/MDAnalysisTests/core/test_atomgroup.py
> <https://github.com/MDAnalysis/mdanalysis/pull/3188#discussion_r603226294>
> :
>
> > +class TestAtomGroupSort(object):
> +    """Tests the AtomGroup.sort attribute"""
> +
> +    @pytest.fixture()
> +    def u(self):
> +        return mda.Universe(PDB_helix)
> +
> +    @pytest.fixture()
> +    def ag(self, u):
> +        return u.atoms[[3, 2, 1, 0]]
> +
> +    @pytest.fixture()
> +    def agsort(self, ag):
> +        return ag.sort("ids")
> +
> +    def test_sort(self, agsort):
>
> @IAlibay <https://github.com/IAlibay> sorry to jump the gun a bit here,
> but would we want to restrict to a list of sortable attributes or check
> there is no collision (scales badly)?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3188#discussion_r603226294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGSGB5OMYH5AF62CXV6CALTGBSEXANCNFSM4ZW3NU2Q>
> .
>

1ut commented:
I assume no one wants to sort atomgroup by characters like names of residues. I'll add lines to check if all elements of the given array are numeric if needed.

IAlibay commented:
> I assume no one wants to sort atomgroup by characters like names of residues. I'll add lines to check if all elements of the given array are numeric if needed.

There might be a rare use case, but I agree it's probably not happening.

Imho it might be safer to only allow for cases where we can predict the behaviour. @richardjgowers thoughts on this? The alternative here is just to consider the atomgroup as a numpy array-esque class, and go with the philosophy that all we are doing is just replicating upstream numpy sort behaviour.

@hmacdope re - collisions, that's part of the reason I brought it up, i.e. can we see what happens in those cases (which is essentially the `masses` case). If we think it's too unpredictable to have collisions, we could restrict the attributes to unique ones (i.e. ix and ids), but personally I'm in favour of going "if it replicates numpy behaviour, then it's fine".

richardjgowers commented:
Nothing wrong with sorting strings? No need to complicate this; if the user
asks for sorted by resname then we can treat them like adults and sort by
resname rather than springing an error on them?

On Mon, Mar 29, 2021 at 14:30, Irfan Alibay ***@***.***>
wrote:

> I assume no one wants to sort atomgroup by characters like names of
> residues. I'll add lines to check if all elements of the given array are
> numeric if needed.
>
> There might be a rare use case, but I agree it's probably not happening.
>
> Imho it might be safer to only allow for cases where we can predict the
> behaviour. @richardjgowers <https://github.com/richardjgowers> thoughts
> on this? The alternative here is just to consider the atomgroup as a numpy
> array-esque class, and go with the philosophy that all we are doing is just
> replicating upstream numpy sort behaviour.
>
> @hmacdope <https://github.com/hmacdope> re - collisions, that's part of
> the reason I brought it up, i.e. can we see what happens in those cases
> (which is essentially the masses case). If we think it's too
> unpredictable to have collisions, we could restrict the attributes to
> unique ones (i.e. ix and ids), but personally I'm in favour of going "if it
> replicates numpy behaviour, then it's fine".
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3188#issuecomment-809378273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGSGB6RLQBFNAI55BLHS4DTGCFIXANCNFSM4ZW3NU2Q>
> .
>

IAlibay commented:
> Nothing wrong with sorting strings? No need to complicate this; if the user asks for sorted by resname then we can treat them like adults and sort by resname rather than springing an error on them?

Makes sense, let's just go with a simple sort then, unless @hmacdope has a specific reason against it.

richardjgowers commented:
@1ut I've noticed you've opened a new PR #3195 - are you going to finish this or otherwise we can close this?

1ut commented:
@richardjgowers 
I didnt intend to give up this issue. Should I close PR #3195 ?

1ut commented:
I noticed that argsort doesn't work properly to sort atomgroup when given multiple dimensional array, so I made the argument "keyfunc" that receives a function to process multiple dimension array that the attribute returns and to return 1D array as a key for sorting.

1ut commented:
@IAlibay 
> Apologies, I'm likely missing something obvious here. Do we have a use case for sorting on non-1D arrays?
Atomgroup.positions returns 2D arrays. When given a key "positions", atomgroup.sort of previous code didn't raise any error and returned atomgroup with destructed .poisitions. 
> Similarly, idx (as set in your code), is a Universe attribute right? Is there a normal case for this being a 0 dimension numpy array?
No. So I removed the lines for checking if idx is 0 dimension array. 
> I agree that there should be a check for non-"number of atoms" sized attributes (off the top of my head, technically someone could try to pass something like "bonds" here and the whole thing would throw (at the moment) an AttributeError). We probably have to be a bit more general about our checking here.
I added line to check if number of atoms and idx have a same size.

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+        """
+        idx = getattr(self.atoms, key)
+        if len(idx) != len(self.atoms):
+            raise ValueError("The array returned by the attribute '{}' "
```
Please add a test to cover this.

1ut replied:
Added test for it. 
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/26b377323c113b0343517ac5473c573d381b60df

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+                             "atoms".format(key))
+        if idx.ndim == 1:
+            order = np.argsort(idx, kind='stable')
+        elif idx.ndim > 1:
```
I'm not fully convinced this is in the scope of what we want to cover with this function, but I'll defer to @richardjgowers' opinion though.

1ut replied:
I thought someone may want to sort by extracts of multiple dimensional arrays like the x-axis of positions. However, I also feel it should just raise an error when `idx.ndim` is not equal to 1.

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+            >>> from MDAnalysisTests.datafiles import PDB_small
+            >>> u = mda.Universe(PDB_small)
+            >>> ag = u.atoms[[3, 2, 1, 0]]
+            >>> ag.ids
```
In my opinion, the inclusion of both `ids` and `ix` in this docstring example is confusing for users. It feels like it hints that the sort method sorts by `ids` because that's the attribute you are showing off first. For the sake of simplicity, I would suggest just having `ag.ix`.

1ut replied:
Fixed.
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/a577298b4985b533233bc52a3a7bf75323055312

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+                             "must have a same length as the number of "
+                             "atoms".format(key))
+        if idx.ndim == 1:
+            order = np.argsort(idx, kind='stable')
```
The `kind` type will need to be documented in the docstring, probably as a Note. Some advanced users might need to know what kind of sort behaviour they should be expecting.

1ut replied:
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/3582d3a9768d20d2414b1d63a4fea2b7e60b2ea0

Added a description for the kind type.

IAlibay commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -1674,3 +1674,70 @@ def test_partial_timestep(self, universe):
+
+    def test_sort_position(self, ag):
+        ag.positions = (-np.arange(21)).reshape(7, 3)
+        with pytest.raises(ValueError):
```
Can we have some `match` arguments to raises here so we check that we are picking up the right error?

1ut replied:
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/f6146d90640eb77d37d2e1090b52a82b1f1d5e47

Added match arguments.

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+                                 "'keyfunc':{} doesn't return 1D array."
+                                 .format(keyfunc))
+            order = np.argsort(sortkeys, kind='stable')
+        agsorted = self.atoms[order]
```
nitpicking, but this seems a bit unnecessary if the next line is just to return this. It's probably understandable enough to just return `self.atoms[order]`?

1ut replied:
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/2f1577bfae5a9a1b166d43e7dede359059c9a7de

Fixed.

IAlibay reviewed (changes requested):
Thanks @1ut, just a couple of comments to address.

IAlibay commented:
I've fixed the merge conflicts so CI can run.

@hmacdope and @richardjgowers can you weigh in on the "sorting > 1 D arrays"?

richardjgowers reviewed (approved):
Lgtm

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+        keys: str
+            The name for attribute of Atomgroup. (e.g. "ids", "ix". default="ix".)
+        keyfunc: function
+            A function that returns 1 dimension array, a key for sorting the
+            atomgroup from multiple dimension array that the attribute returns.
```
```suggestion
            A function to convert multidimensional arrays to a single
            dimension. This 1D array will be used as the sort key and
            is required when sorting with an ``AtomGroup`` attribute
            key which has multiple dimensions. Note: this argument
            is ignored when the attribute is one dimensional.
```

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+        idx = getattr(self.atoms, key)
+        if len(idx) != len(self.atoms):
+            raise ValueError("The array returned by the attribute '{}' "
+                             "must have a same length as the number of "
+                             "atoms".format(key))
```
```suggestion
                             "must have the same length as the number of "
                             "atoms in the input AtomGroup".format(key))
```

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+                                "dimension array.".format(key))
+            sortkeys = keyfunc(idx)
+            if sortkeys.ndim != 1:
+                raise ValueError("The function you assigned to the argument "
```
```suggestion
                raise ValueError("The function assigned to the argument "
```

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,66 @@ def write(self, filename=None, file_format=None,
+            order = np.argsort(idx, kind='stable')
+        elif idx.ndim > 1:
+            if keyfunc is None:
+                raise NameError("You have to assign the argument 'keyfunc' "
```
I think this message needs reversing in order to be more user friendly.

"The {} attribute returns a multidimensional array. In order to sort it, a function returning a 1D array (to be used as the sort key) must be passed to the `keyfunc` argument"

1ut replied:
Fixed. 

https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/e0bef58681c2fba19490ceb1d964740f60518e2d
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/338417e2ea1c71d908d22037412a8d7d5c014b72

IAlibay reviewed (changes requested):
Nearly there, just some comments on the documentation text.

1ut commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -1674,3 +1674,74 @@ def test_partial_timestep(self, universe):
+
+    def test_sort_position(self, ag):
+        ag.positions = (-np.arange(21)).reshape(7, 3)
+        with pytest.raises(ValueError, match=r"The array returned by the "
```
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/52b677a50c2977a0df07f40ba462c230f01974fb

Fixed.

1ut commented:
@IAlibay 
Thank you for your kind suggestions for documentations. I committed all changes.

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,82 @@ def write(self, filename=None, file_format=None,
         raise ValueError("No writer found for format: {}".format(filename))
 
+    def sort(self, key='ix', keyfunc=None):
+        """Returns a sorted ``AtomGroup`` using a specified attribute as \
```
Unecessary `\`?

1ut replied:
Fixed.
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/33776558e9dc6ce659dfc57537d98aeb80209ea7

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,82 @@ def write(self, filename=None, file_format=None,
+            >>> ag = ag.sort()
+            >>> ag.ix
+            array([0 1 2 3])
+            >>> ag.positions
```
Awesome thanks for adding this example too!

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,82 @@ def write(self, filename=None, file_format=None,
+
+        Returns
+        ----------
+        Group
```
Looking at the docs (https://mdanalysis--3188.org.readthedocs.build/en/3188/documentation_pages/core/groups.html?highlight=sort#MDAnalysis.core.groups.AtomGroup.sort), this sets the "Return type" to "Group". We probably want to have this be a :class:`AtomGroup`?

```suggestion
        :class:`AtomGroup`
```

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,82 @@ def write(self, filename=None, file_format=None,
+        This uses a stable sort as implemented by
+        `numpy.argsort(kind='stable')`.
+
+        .. versionadded:: 2.0.0
```
Looking at the docs, the versionadded is being placed inside the `Note` box. We'll need two blank lines here.
```suggestion

        .. versionadded:: 2.0.0
```

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3324,6 +3324,82 @@ def write(self, filename=None, file_format=None,
+            sortkeys = keyfunc(idx)
+            if sortkeys.ndim != 1:
+                raise ValueError("The function assigned to the argument "
+                                 "'keyfunc':{} doesn't return 1D array."
```
```suggestion
                                 "'keyfunc':{} doesn't return a 1D array."
```

IAlibay commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -1674,3 +1674,82 @@ def test_partial_timestep(self, universe):
+
+    def test_sort_positions_2D(self, ag):
+        with pytest.raises(ValueError, match=r"The function assigned to"
+                           ".*"):
```
you shouldn't need the `.*` here (or above/below), it _should_ by default by a substring of the error message being raised.

1ut replied:
https://github.com/MDAnalysis/mdanalysis/pull/3188/commits/c2f734260ee05bd387572144b5a0a15463cf53e0

Fixed.

IAlibay reviewed (changes requested):
Nearly there, just a couple of small things, mainly docstring changes to make sure the generated docs look nice.

IAlibay reviewed (approved):
Assume CI returns green, lgtm!

hmacdope reviewed (approved):
great work @1ut, lgtm also 👍
